numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -326,41 +328,38 @@ PyArray_AdaptFlexibleDType(PyObject *data_obj, PyArray_Descr *data_dtype,
              */
             PyErr_SetString(PyExc_TypeError,
                     "don't know how to adapt flex dtype");
-            *flex_dtype = NULL;
-            return;
+            Py_DECREF(retval);
+            return NULL;
         }
     }
     /* Flexible type with generic time unit that adapts */
     else if (flex_type_num == NPY_DATETIME ||
                 flex_type_num == NPY_TIMEDELTA) {
-        meta = get_datetime_metadata_from_dtype(*flex_dtype);
+        meta = get_datetime_metadata_from_dtype(flex_dtype);
```

Comment:
Look like there should be a DECREF right aftr this call. Also removes the need for the decref's later.

Reply:
adopted

Reply:
I am confused what I had thought here, but the code looks right...

Reply:
This was the segfault. When removing the `retval = flex_dtype` at the top of the function and moving it to an else clause at the end, I didn't take into account the implied `else` cases in this block. You had suggested moving the `Py_DECREF(flex_type)` out to here, but instead I moved `retval = flex_type` here and left the `Py_DECREF`s in place.